pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,298 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
+		return scaleDownLimited(f, gameServerLister, nodeCounts, key, isCounter, replicas,
+			aggCapacity, maxCapacity)
+	case 0:
+		return replicas, false, nil
+	}
+
+	return 0, false, errors.Errorf("cannot scale due to error in scaleLimited function")
+}
+
+// scaleUp scales up for either Integer or Percentage Buffer.
+func scaleUp(replicas int32, capacity, count, aggCapacity, availableCapacity, maxCapacity,
+	minCapacity, buffer int64) (int32, bool, error) {
+
+	// How much capacity is gained by adding one more replica to the fleet.
+	replicaCapacity := capacity - count
+	if replicaCapacity <= 0 {
+		return 0, false, errors.Errorf("cannot scale up as adding additional replicas does not increase available Capacity")
+	}
+
+	additionalReplicas := int32(math.Ceil((float64(buffer) - float64(availableCapacity)) / float64(replicaCapacity)))
+
+	// Check to make sure we're not limited (over Max Capacity)
+	limited, _ := isLimited(aggCapacity+(int64(additionalReplicas)*capacity), minCapacity, maxCapacity)
+	if limited {
+		additionalReplicas = int32((maxCapacity - aggCapacity) / capacity)
+	}
+
+	return replicas + additionalReplicas, limited, nil
+}
+
+// scaleDown scales down for either Integer or Percentage Buffer.
+func scaleDown(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount, key string, isCounter bool, replicas int32,
+	availableCapacity, aggCount, aggCapacity, minCapacity, buffer int64) (int32, bool, error) {
+	// Exit early if we're already at MinCapacity or DesiredCapacity to avoid calling
+	// getSortedGameServers if unnecessary
+	if aggCapacity == minCapacity {
+		return replicas, true, nil
+	}
+
+	if availableCapacity == buffer {
+		return replicas, false, nil
+	}
+
+	// We first need to get the individual game servers in order of deletion on scale down, as any
+	// game server may have a unique value for counts and / or capacity.
+	gameServers, err := getSortedGameServers(f, gameServerLister, nodeCounts)
+	if err != nil {
+		return 0, false, err
+	}
+
+	// "Remove" one game server at a time in order of potential deletion. (Not actually removed here,
+	// that's done later, if possible, by the fleetautoscaler controller.)
+	for _, gs := range gameServers {
+		replicas--
+		switch isCounter {
+		case true:
+			if counter, ok := gs.Status.Counters[key]; ok {
+				aggCount -= counter.Count
+				aggCapacity -= counter.Capacity
+			} else {
+				continue
+			}
+		case false:
+			if list, ok := gs.Status.Lists[key]; ok {
+				aggCount -= int64(len(list.Values))
+				aggCapacity -= list.Capacity
+			} else {
+				continue
+			}
+		}
+		availableCapacity = aggCapacity - aggCount
+		// Check if we're Limited (Below MinCapacity)
+		if aggCapacity < minCapacity {
+			return replicas + 1, true, nil
+		}
+		// Check if we're at our desired Buffer
+		if availableCapacity == buffer {
+			return replicas, false, nil
+		}
+		// Check if we're at Limited
+		if aggCapacity == minCapacity {
```

Comment:
I think this check should happen before line 513, to indicate it's scaleLimited in the case where `aggCapacity == minCapacity` and `availableCapacity == buffer` are true at the same time.

Reply:
I interpret "limited" to mean that we could not reach our desired buffer because we hit min or max capacity, which is why I first check for the buffer. A different interpretation would be that "limited" means that we are at min or max capacity (regardless of the desired capacity).

Based on the comment `// scalingLimited indicates that the calculated scale would be above or below the range defined by MinReplicas and MaxReplicas` in /pkg/fleetautoscalers/controller.go#L395 I'm leaning towards the first interpretation.

Reply:
Make sense!